pub/helpers/bin/infer_sdk_versions.dart

```diff
@@ -0,0 +1,162 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:collection/collection.dart';
+import 'package:http/http.dart';
+import 'package:http/retry.dart';
+import 'package:yaml/yaml.dart';
+import 'package:pub_semver/pub_semver.dart';
+import 'package:path/path.dart' as p;
+
+Never fail(String message) {
+  stderr.writeln(message);
+  exit(-1);
+}
+
+final client = RetryClient(Client());
+
+ArgResults parseArgs(List<String> args) {
+  final argParser = ArgParser()
+    ..addOption(
+      'directory',
+      abbr: 'C',
+      defaultsTo: '.',
+      help: 'The directory containing the pubspec.yaml of the package.',
+    )
+    ..addOption('flutter-releases-url',
+        help:
+            'The url to retrieve the list of available flutter releases from.')
+    ..addFlag('help', help: 'Display the usage message.');
+  final results;
+  try {
+    results = argParser.parse(args);
+    if (results['help'] as bool) {
+      stdout.writeln(
+          'Infers the newest available flutter sdk to use for a package.');
+      stdout.writeln(argParser.usage);
+      exit(0);
+    }
+    return results;
+  } on FormatException catch (e) {
+    stderr.writeln(e.message);
+    stderr.writeln(argParser.usage);
+    exit(-1);
+  }
+}
+
+Map<String, VersionConstraint> parseSdkConstraints(dynamic pubspec) {
+  final dartConstraint =
+      VersionConstraint.parse(pubspec['environment']?['sdk'] ?? 'any');
+  final flutterConstraint =
+      VersionConstraint.parse(pubspec['environment']?['flutter'] ?? 'any');
+  return {
+    'dart': dartConstraint,
+    'flutter': flutterConstraint,
+  };
+}
+
+Future<void> main(List<String> args) async {
+  try {
+    final argResults = parseArgs(args);
+    final flutterReleases = await retrieveFlutterReleases(
+        argResults['flutter-releases-url'] ?? flutterReleasesUrl);
+
+    final pubspecPath = p.join(argResults['directory'], 'pubspec.yaml');
+    final pubspec = loadYaml(File(pubspecPath).readAsStringSync(),
+        sourceUrl: Uri.file(pubspecPath));
```

Comment:
Inferring is a good start, but dependabot should allow pointing to an exact git-ref (`tag`, `sha1`).

Reply:
Hmm - do you have an idea for a good place to put this configuration?

What are scenarios where you would need this?

Reply:
Scenario: We use [`flutterw`](https://github.com/passsy/flutter_wrapper) to pin the flutter version. Others use [`fvm`](https://github.com/fluttertools/fvm) or custom solutions.

Tree options pop into my mind:
- Lookup: dependabot checks the version of those pinning solutions (i.e. check git submodule sha1 or a configuration file)
- Add a flutter version to .github/dependabot.yaml
- Define a new `dependabot.flutter_version` property to `pubspec.yaml` of each package and read it

Reply:
cc @jonasfj what do you think?

Reply:
I think that it sufficient that we allow pinning the Flutter SDK constraint in `environment.flutter`.

Generally, the `dart pub` client will ignore the upper-bound in `environment.flutter`, as a result of a decision made in Flutter 2 that Flutter wouldn't break backwards compatibility.

But in dependabot, I think we made it respect the upper-bound, thus, the `environment.flutter` constraint can be used to pin a Flutter version. You can't pin to arbitrary sha or tag, but you can pin to stable and beta releases of Flutter.